include/revng/Model/Binary.h

```diff
@@ -282,15 +283,51 @@ class model::Binary : public model::generated::Binary {
   bool verifyTypeDefinitions(bool Assert) const debug_function;
   bool verifyTypeDefinitions() const debug_function;
 
-  bool verifyGlobalNamespace(VerifyHelper &VH) const;
-  bool verifyGlobalNamespace(bool Assert) const debug_function;
-  bool verifyGlobalNamespace() const debug_function;
-
 public:
   void dumpTypeGraph(const char *Path) const debug_function;
 
 public:
   MetaAddressRangeSet executableRanges() const;
+
+private:
+  // It's defined here instead of the TTG because we don't really want it to
+  // ever get serialized.
+  mutable std::optional<model::TypeOfGlobalNameCache> TheGlobalNameCache = {};
+  std::optional<model::TypeOfGlobalNameCache>
+  buildGlobalNameCache(VerifyHelper &VH) const;
+
+public:
+  const model::TypeOfGlobalNameCache *GlobalNameCache(VerifyHelper &VH) const {
+    if (not TheGlobalNameCache) {
+      TheGlobalNameCache = buildGlobalNameCache(VH);
+      if (not TheGlobalNameCache)
+        return nullptr;
+    }
+
+    return &TheGlobalNameCache.value();
+  }
+  const model::TypeOfGlobalNameCache *
+  GlobalNameCache(bool Assert) const debug_function {
+    VerifyHelper VH(Assert);
+    return GlobalNameCache(VH);
+  }
+  const model::TypeOfGlobalNameCache *GlobalNameCache() const debug_function {
+    return GlobalNameCache(false);
+  }
+
+  void discardGlobalNameCache() const { TheGlobalNameCache = std::nullopt; }
+
+public:
+  inline std::string
+  getLLVMFunctionName(const model::Function &Function) const {
```

Comment:
1. `inline` is not necessary
2. In LLVM we don't have all the issues we have in C, I think we can just use `CustomName` directly, if available.

Reply:
1. A left-over, my bad. This was a free function and I didn't pay attention.
2. Yes, but if we do that, we aren't going to get `_function_0x_$SOMEWHERE_Code_$SOMETHING` labeled correctly. And I'd rather keep that logic in one place.

But considering this is going to become a member of the new `Namer` thing, I think it will work out nicely.

Reply:
1. Done.

Reply:
1. Done.